README
Chaton (pronounced like [sha-ton]) is a comet-based webchat server.
It is a collection of CGI scripts, daemon programs and other scripts.

Prerequisites:

- Make sure you can run daemon processes on your server.
  It might be difficult if you're using a shared hosting
  service.

- It is also desirable that you can make your own crontab.

- The server should also run httpd that is capable to run
  CGI.  The server should be configured so that you can
  do SetHandler in .htaccess file.

- Gauche 0.8.14 or later.  You can install it under
  your home directory.


Files:

  The following files are scripts that cooperates to make Chaton
  work:

    chaton-entry    This is a CGI script invoked when the room
          is accessed via web browsers.  This script creates the
          "outer frame" of the chat room page, and make the "inner
          frame" connect to the Comet server.

    chaton-viewer   This is the Comet server responsible to
          distribute room contents.   It must run as a daemon.
          Currently you have to run one daemon process per
          one chatroom.  The actual installed script is named
          'chaton-viewer-$ROOM', where $ROOM is the room name.

    chaton-poster   This is a CGI script that handles message
          posting.  The actual installed script is named
          'chaton-poster-$ROOM'.

    chaton-browser  This is a CGI script to browse message archives.
          When installed, this script is renamed to 'a', for the sake
          of shorter permalinks.

    chaton-archiver  This is a script file that archives message
          logs.  This script must be run daily.

    chaton-badge    This is a CGI script that generates "badge"
          of the room.  When installed, this script becomes 'b',
          for conciseness.

    chaton-apilogin  This is a CGI script where the client programs
          first contact through API.  When installed, this script
          becomes 'apilogin'.


 The following files consist the client-side contents:

    chaton.js   The javascript that implements client-side operations.
          The same file is read from both the outer content and
          and from the content inside the iframe.
          Upon installation, this file is renamed to 'chaton.XXXXXX.js'
          where XXXXXX is a unique string, to avoid caching by
          browsers.

    chaton.css   The stylesheet.  Used by both the outer content
          and the content inside the iframe.

  Additionally, 'chaton.scm' implements common funcionalities used
  among server-side scripts, and 'htaccess' is the skeleton of
  '.htaccess' file for each room.


Directory structure:

  $BINDIR        A directory where server-side programs (chaton-viewer-*
                 and chaton-archiver-*) will go.  You may want this
                 directory to be in your PATH for the convenience,
                 although not necessarily.

  $DATADIR       A directory where the chat messages and other stuff
                 are stored.  Recommended not to be under the directory
                 visible via httpd.
                 Each chatroom must have distinct $DATADIR.

    $DATADIR/data   Subdirectory where message data goes into.  CGI
                 scripts need to write into this directory.

    $DATADIR/data/current.dat   Current message data.  Typically it
                 contains around one-day worth of postings.   The
                 chaton-archiver script removes old messages from this
                 and stores them to archived data.

    $DATADIR/data/YYYY/MM/DD.dat  Archived data.   Created by
                 chaton-archiver script.   Chaton-archiver also sets
                 the today's data file as a symlink to the current.dat
                 file.

    $DATADIR/logs/comet.log  Access log file from the Comet server
                 (chaton-viewer-*).

    $DATADIR/logs/archive.log  Log from the archiver.


  $HTDOCSDIR     A directory where the content accessed by browsers
                 will go.   Each chatroom must have distinct $HTDOCSDIR.
                 The httpd should honor .htaccess in this directory.

    $HTDOCSDIR/.htaccess     Some DirectoryIndex and SetHandler directives.

    $HTDOCSDIR/entry         The main entry CGI.  We use DirectoryIndex so
                 that this script is kicked when the client access
                 this directory.

    $HTDOCSDIR/chaton.css, $HTDOCSDIR/chaton.XXXXXX.js
                 CSS and javascript files.

    $HTDOCSDIR/chaton-poster-$ROOM   A CGI script to post the message.

    $HTDOCSDIR/a   A CGI script to browse the archived messages.  From
                 the web browsers, archived messages are accessed
                 as 'http://yourdomain.example.com/.../$ROOM/a/YYYY/MM/DD'

    $HTDOCSDIR/var           A directory holds some transient data.  This
                 directory must be writable both by CGI scripts and
                 the user running viewer and archiver.

    $HTDOCSDIR/var/status.{js,scm}  The comet server periodically dumps
                 the room status into these files.  You can poll these files
                 to find out the current status of the room.
                 See http://chaton.practical-scheme.net/doc?Retrieving%20Room%20Status
                 for the details.

    $HTDOCSDIR/var/index.rdf
    $HTDOCSDIR/var/seq       These are written by chaton-poster-$ROOM script.
                 The index.rdf serves RSS.  The seq file is a text file contains
                 a number; they can be read by the clients to show # of unread
                 messages.

Build a chat room:

  To build a chat room, first you prepare configuration file.
  Copy 'sample.conf' and follow the directions.

  The 'sample.conf' file includes two auxiliary files, 'conf/site.conf'
  and 'conf/appearance.conf'.   You MUST edit 'conf/site.conf' to reflect
  your site settings; the default values are useless to you.  On the
  other hand, you can leave 'conf/appearance.conf' as is if you don't
  need to change it.

  Other than those included .conf files, each chat room MUST have
  distinct copy of configuration file.  Especially, make sure
  that the room name, comet port number, url path, htdocs dir and
  data dir are all unique for each chat room.

  Once you prepare conf file, run build-site script:

    gosh ./build-site yourroom.conf

  This copies necessary files, with modifying them according to
  the configuration settings.

  For your convenience, you can list multiple conf files in the
  file 'rooms', one per line.  Then simply the following command
  runs build-site for every room you have.

    make install

  If you run cgi scripts in a different user privildge than yourself
  (that is, you're not using suexec), you need to adjust permissons
  of some directories.  See the above directory structure descriptions.
  The recommended way is to use common group permissoins and make those
  directories group writable.
  It'd be nicer to do so automagically by build-site script, but not yet.


Run a chat room:

  You have to run chaton-viewer-$ROOM as a background process.

  You also have to set up chaton-archiver-$ROOM to be executed
  once a day, probably using cron.



ChangeLog
2018-12-03  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter (kick-reply-watcher!): Remove
	  using twitter user stream.  For now, twitter feed is only one
	  direction (chaton -> twitter).

2016-03-26  Shiro Kawai  <shiro@acm.org>

	* chaton-entry, chaton.js: Add "'Return' key to post" checkbox.

2016-02-12  Shiro Kawai  <shiro@acm.org>

	* chaton.js: Shift+ENTER in textarea inserts newline instead
	  of posting.

2015-10-24  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter (feed-to-twitter): Now they count
	  url as 23 character long.

2014-10-19  Shiro Kawai  <shiro@acm.org>

	* chaton-poster (make-entry): Make sure invalid Char isn't written
	  into datafile.

	* chaton.scm (chaton-render-rss-1): A quick fix not to include
	  control characters in rss feed outside of CDATA.  Proper fix is
	  still required to filter such characters at the posting.

2014-03-09  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter (feed-to-twitter): Fixed the post
	  length counting, w.r.t. twitter's t.co link shortening.
	  Since twitter shortens urls no matter how long the original ones
	  are, we no longer need to use bit.ly on our end.  Need to remove
	  it in future.

2012-06-26  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter (forward-from-twitter): Filter out
	  retweets.

2011-03-20  Shiro Kawai  <shiro@acm.org>

	* chaton-viewer (get-header etc.): Support HTTP/1.0 request as well,
	  for some proxy servers (e.g. Squid) may send it.

2010-08-31  Shiro Kawai  <shiro@acm.org>

	* chaton.scm (chaton-alist->stree): Render nested lists in the value
	  of the map as an array in json mode.
	* chaton-viewer: Changed API to recognize "s=0" as a json mode which
	  returns the same info as Sexpr mode.  The default (missing 's')
	  behavior is a special format just for web interface.
	  THIS IS AN INCOMPATIBLE CHANGE: It breaks a client that sends
	  's' parameter with a value other than 1 and expects Sexpr.
	  Clients using chaton.client module won't break, for the module
	  already sends '1'.

	* chaton-apilogin: Allow returning result in json if "s=0" parameter
	  is given.
	* tests/apilogin.scm, tests/Makefile: Added apilogin test.

2010-06-07  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter: Switched from Basic authentication
	  to OAuth to access Twitter.  Requires Gauche-net-twitter package.

2010-04-12  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter (kick-reply-watcher!): run a watchdog
	  thread to prevent twitter api call from being wedged.

2010-03-26  Shiro Kawai  <shiro@acm.org>

	* chaton-archiver (filter-entry): Last fix was incorrect.  Fixed again.

2010-03-25  Shiro Kawai  <shiro@acm.org>

	* chaton-archiver (filter-entry): Fixed a bug that truncated current.dat
	  when it has less than MIN-ENTRIES entries and all entries are
	  before cut-off date.

2009-11-13  Shiro Kawai  <shiro@acm.org>

	* client/chaton/client.scm (make-handler): When an unexpected error
	  occurs within observer loop, restart the thread loop instead of
	  exitting the thread.  Unexpected error can occur time to time,
	  e.g. temporary network failures.

2009-10-26  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter (kick-reply-watcher!): Insert delay
	  before resuming watcher thread loop after capturing an error.
	  Without the delay, the thread can get into a busy loop if a
	  malfunction persists for some time (e.g. network failure).

2009-10-19  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter (make-observer): return #f when
	  observer captures an error, so that observer thread won't terminate
	  when twitter or bit.ly access fail.

2009-10-16  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter (forward-from-twitter): Only
	  forward messages from twitter by the users who follow the bridged
	  chaton room.  This eliminates bouncing of multipost messages,
	  pikcing up unintended retweets, etc.

2009-10-13  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter (kick-reply-watcher!): fixed a bug
	  in the handler clause of guard that caused to terminate watcher
	  thread.

2009-10-10  Shiro Kawai  <shiro@acm.org>

	* client/examples/chaton-twitter: Added a sample chaton-twitter bridge.

	* client/chaton/client.scm (chaton-permalink): Added utility function
	  to derive permalink from <chaton-client> and timestamp.

2009-09-21  Shiro Kawai  <shiro@acm.org>

	* chaton.scm (chaton-render-rss-1): Use html-escape-string instead
	  of safe-text for escaping TITLE elements of RSS, since safe-text
	  may insert A elements which aren't allowed in TITLE.

2009-07-29  Shiro Kawai  <shiro@acm.org>

	* chaton.scm (+show-stack-trace+): missing '@@' markers.
	  (chaton-render-from-file): enable to switch the order of entries
	  by a keyword argument newest-first.  It is better to reverse the
	  order for RSS.

2009-07-26  Shiro Kawai  <shiro@acm.org>

	* chaton.scm (chaton-render-rss-1): fix title.

	* chaton-entry (the-page), chaton.scm (+index.rdf+): Changed the
	  location of index.rdf under 'var' subdirectory; the room directory
	  may be unwritable.
	* chaton-poster (update-rss): Use extended with-output-to-file
	  to avoid race condition

	* chaton.scm (chaton-render-html-1, chaton-render-rss-1): Refactored
	  common routines.

	* chaton-poster (poster-main), chaton.scm (+show-stack-trace+),
	  sample.conf (show-stack-trace-on-error): Added configuration
	  switch to customize whether stack trace should be shown on
	  error.

2009-07-26  Kazuki Tsujimoto <kazuki@callcc.net>

	* cahton-poster (poster-main): Show stack trace when an error
	  occured.

	* chaton-poster (update-rss): Added to support RSS.
	* chaton-entry, chaton.scm: Changed accordingly.

2009-06-13  Shiro Kawai  <shiro@acm.org>

	* chaton.scm (render-url): Allow '-' in the youtube vid.

	* chaton-archiver (mklink): Ensure directory before creating the
	  symlink (only becomes a problem when archiver is invoked for the

tests/lib.scm
;;
;; basic tests for chaton.scm
;;

(use gauche.test)

(test-start "chaton.scm")
(load "../chaton.scm")
(test-module 'chaton)

(test-end)


client/test.scm
(use gauche.test)
(add-load-path ".")

(test-start "chaton client")

(test-section "chaton.client")
(use chaton.client)
(test-module 'chaton.client)

(test-end)


tests/badge.scm
;;
;; basic tests for badge
;;

(use gauche.test)

(add-load-path "..")

(test-start "badge")
(load "../chaton-badge")
(define main #f)                        ;prevent execution of 'main'
(test-module 'chaton.badge)

(test-end)


chaton.js
//
// NB: This script is read in both the main html and inner frame html
//

//=================================================================
// Functions for outer html
//

// Post interface ----------------------------------

function post() {
  setNickCookie($F('post-remember'));
  setReturnPostCookie($F('post-by-return'));
  if ($F('post-nick') == '' || $F('post-text') == '') return;
  disablePost();
  new Ajax.Request("@@httpd-url@@@@url-path@@@@cgi-script@@",
    {
      parameters : {
        nick: $F('post-nick'), 
	text: $F('post-text')
      },
      onSuccess : function (t) { enablePost(true); },
      onFailure : function (t) { enablePost(false); },
      onException : function (r, e) { enablePost(false); }
    });
}

function textKey(e) {
  if ($('post-text').disabled) return;
  var key = (e.which || e.keyCode);
  if (key == Event.KEY_RETURN) {
    if (e.shiftKey || $('post-by-return').checked == false) {
      $('post-text').insert('<br/>');
    } else {
      post();
    }
  }
}

function disablePost() {
  $('post-submit').disabled = true;
  $('post-nick').disabled = true;
  $('post-text').disabled = true;
  $('post-form').disabled = true;
}

function enablePost(clearp) {
  $('post-submit').disabled = false;
  $('post-nick').disabled = false;
  $('post-text').disabled = false;
  $('post-form').disabled = false;
  if (clearp) { $('post-text').clear(); }
  $('post-submit').focus();
  $('post-text').focus();
}

function initByCookie() {
    var cookies = document.cookie.split(';');
    var len = cookies.length;
    for (var i = 0; i < len; i++) {
        var sc = cookies[i].strip();
        if (sc.startsWith('chaton-nickname=')) {
            var nick = sc.substring('chaton-nickname='.length, sc.length);
            $('post-nick').value = unescape(nick);
            $('post-remember').checked = true;
        }
        if (sc == 'chaton-no-post-by-return=1') {
            $('post-by-return').checked = false;
        } else {
            $('post-by-return').checked = true;
        }
    }
}

function setNickCookie(set) {
    if (set) {
        document.cookie = 'chaton-nickname=' + escape($F('post-nick'))
            + ';expires=Tue, 19 Jan 2038 00:00:00 GMT'
            + ';path=@@cookie-path@@';
    } else {
        document.cookie = 'chaton-nickname='
            + ';expires=Thu, 01-Jan-1970 00:00:01 GMT'
            + ';path=@@cookie-path@@';
    }
}

function setReturnPostCookie(set) {
    if (set) {
        document.cookie = 'chaton-no-post-by-return='
            + ';expires=Tue, 19 Jan 2038 00:00:00 GMT'
            + ';path=@@cookie-path@@';
    } else {
        document.cookie = 'chaton-no-post-by-return=1'
            + ';expires=Thu, 01-Jan-1970 00:00:01 GMT'
            + ';path=@@cookie-path@@';
    }
}
   
// Sequence count monitor -----------------------------

var messageMonitorRunning = false;
var messageMonitorContinue = false;
var currentMessageNum = -1;
var viewedMessageNum = -1;

function setTitle() {
  if (messageMonitorContinue) {
    var num = currentMessageNum - viewedMessageNum;
    if (num > 0) {
      window.document.title = '[' + num + '] Chaton @@room-name@@';
      return;
    }
  }
  window.document.title = 'Chaton @@room-name@@';
}

function messageMonitorRun() {
  messageMonitorContinue = true;
  if (!messageMonitorRunning) {
    messageMonitorRunning = true;
    currentMessageNum = viewedMessageNum = -1;
    fetchMessageCount();
  }
}

function messageMonitorStop() {
  messageMonitorContinue = false;
  currentMessageNum = viewedMessageNum = -1;
  setTitle();
}

function fetchMessageCount() {
  if (!messageMonitorContinue) {
    messageMonitorRunning = false;
    return;
  }
  new Ajax.Request("@@httpd-url@@@@url-path@@var/seq",
    {
      method: 'get',
      evalJSON: false,
      onSuccess: fetchMessageCountCB,
      onFailure: function (r,e) { messageMonitorStop(); }
    });
}

function fetchMessageCountCB(t) {
  if (!messageMonitorContinue) {
    messageMonitorRunning = false;
    return;
  }
  var cnt = parseInt(t.responseText);
  if (currentMessageNum < 0 || currentMessageNum > cnt) {
    currentMessageNum = viewedMessageNum = cnt;
  } else {
    currentMessageNum = cnt;
  }
  setTitle();
  setTimeout(fetchMessageCount, 15000);
}

// Initialization -------------------------------------
function initMainBody() {
  $('post-text').observe('keypress', textKey);
  $('the-body').onmouseover = function () { messageMonitorStop(); }
  $('the-body').onmouseout  = function () { messageMonitorRun(); }
  initByCookie();
}

//=================================================================
// Functions for inner html
//

var pos = 0;
var seq = 0;
var need_scroll = false;

function fetchContent(cid) {
    seq = (seq+1)%100;
    var ts = ((new Date).getTime()).toString(36) + seq.toString(36);
    new Ajax.Request('/?t=' + ts + '&p=' + pos + '&c=' + cid,
      {
          method: 'get',
          evalJSON: 'force',
          // When Comet server disconnects, Firefox calls onSuccess with
          // emtpy content, while IE7 calls onFailure.
          onSuccess: function(t) {
              var json = t.responseJSON;
              if (!(json && json.ver && typeof(json.cid) == 'number')) {
                  fetchRetry(cid);
              } else {
                  insertContent(json, cid);
              }
          },
          onFailure: function(t) {
              fetchRetry(cid);
          }
      });
    startWatchDog(cid);
}

function fetchRetry(cid) {
    tameWatchDog();
    showStatus('Connection Lost.  Retrying...', 'status-alert');
    setTimeout(function () { resumeFetch(cid); }, 5000 + irandom(10000));
}

function insertContent(json, cid) {
    tameWatchDog();
    if (json.ver != '@@version@@') {
        // The comet server is updated.  We replace the entire document.
        document.location.href = '@@httpd-url@@:@@comet-port@@/';
        return;
    }
    if (json.cid < 0) {
        showStatus('Session Expired.  Please Reload.', 'status-alert');
        return;
    }
    showStatus('Connected ('+json.nc+' user'+(json.nc>1?'s':'')+' chatting)',
               'status-ok');
    need_scroll = true;
    if (json.pos < pos || json.refresh) {
        $('view-pane').update('');
    } else if (!isViewingBottom()) {
        need_scroll = false;
    }
    $('view-pane').insert(json.content);
    var pos_changed = (pos != json.pos);
    pos = json.pos;
    if (need_scroll) scrollToBottom();
    setTimeout(function () { fetchContent(json.cid); },
               pos_changed?irandom(1000):irandom(3000));
}

function resumeFetch(cid) {
    showStatus('Connecting...', 'status-ok');
    fetchContent(cid);
}

function showStatus(text, klass) {
    var st = $('status-line');
    st.removeChild(st.childNodes.item(0));
    st.insert('<span class=\"'+klass+'\">'+text+'</span>');
}

function checkImageSize(img) {
    var img = $(img);
    if (img.width > img.height) {
        if (img.width > @@img-size-limit@@) img.addClassName('wshrunk');
    } else {
        if (img.height > @@img-size-limit@@) img.addClassName('hshrunk');
    }
    img.style.display = 'inline';
    img.removeClassName('hide-while-loading');
    if (need_scroll) scrollToBottom();
}

function scrollToBottom() {
    var sp = $('status-pane');
    if (sp) sp.scrollTo();
}

// Hint from toru@torus.jp
function isViewingBottom() {
    var d = document;
    var de = d.documentElement;
    var w = window;
    var winh = (w.innerHeight
                || d.body.clientHeight       // IE quirk
                || (de && de.clientHeight)   // IE strict
                || 0);
    var winy = (typeof(w.pageYOffset) == 'number'
                ? w.pageYOffset
                : (typeof(d.body.scrollTop) == 'number'
                   ? d.body.scrollTop // IE quirk
                   : ((de && typeof(de.scrollTop) == 'number')
                      ? de.clientHeight // IE strict
                      : 0)));

    if (winh == 0 && winy == 0) return true;
    return winy > getDocumentHeight() - winh - 20;
}

// http://james.padolsey.com/javascript/get-document-height-cross-browser/
function getDocumentHeight() {
    var d = document;
    return Math.max(
        Math.max(d.body.scrollHeight, d.documentElement.scrollHeight),
        Math.max(d.body.offsetHeight, d.documentElement.offsetHeight),
        Math.max(d.body.clientHeight, d.documentElement.clientHeight)
    );
}

function irandom(n) {
    for (;;) {
        var r = Math.floor(Math.random() * n);
        if (r != n) return r;
    }
}

// Watchdog ------------------------------------------
//  Some browsers (e.g. Safari) do not call any callback
//  when server disconnects Ajax client.  As a safety net we set
//  watchdog timer for Comet connection.  In normal circumstances
//  the Comet server replies at most 6 minutes---if we don't hear
//  from the server for 8 minutes, we assume the connection is lost
//  and reload the whole document.

var dog_id = null;

function startWatchDog(cid) {
    if (dog_id) clearTimeout(dog_id);                       // just in case
    dog_id = setTimeout(function () {bark(cid);}, 8*60000); // 8 minutes
}

function bark(cid) {
    // We don't retry fetch, since there's no reliable way to cancel
    // the ongoing Ajax request.  We replace the whole document instead.
    showStatus('Connection Lost.  Retrying...', 'status-alert');
    document.location.href = '@@httpd-url@@:@@comet-port@@/';
}

function tameWatchDog() {
    if (dog_id) {
        clearTimeout(dog_id);
        dog_id = null;
    }
}

// Initialization -------------------------------------
function initViewFrame(cid) {
    setTimeout(function () { fetchContent(cid); }, 1);
}




chaton.scm
;;;
;;; Some common routines for chaton scripts
;;;
(define-module chaton
  (use srfi-1)
  (use srfi-13)
  (use srfi-19)
  (use text.html-lite)
  (use text.tree)
  (use file.util)
  (use util.match)
  (use util.list)
  (use gauche.fcntl)
  (use gauche.sequence)
  (use gauche.experimental.lamb)
  (export chaton-render chaton-read-entries
          chaton-render-from-file
          chaton-render-html-1 chaton-render-rss-1
          chaton-with-shared-locking chaton-with-exclusive-locking

          chaton-alist->stree

          +room-url+ +archive-url+
          
          +datadir+ +current-file+ +sequence-file+
          +last-post-file+ +num-chatters-file+
          
          +logdir+

          +docdir+ +status.js+ +status.scm+ +index.rdf+

          +show-stack-trace+
          
          with-output-to-file))
(select-module chaton)

;;; Some common constants
(define-constant +room-url+    "@@httpd-url@@@@url-path@@")
(define-constant +archive-url+ (build-path +room-url+ "a"))

(define-constant +datadir+ (or (sys-getenv "CHATON_DATADIR")
                               "@@server-data-dir@@data"))
(define-constant +current-file+ (build-path +datadir+ "current.dat"))
(define-constant +sequence-file+ (build-path +datadir+ "sequence"))
(define-constant +last-post-file+ (build-path +datadir+ "last-post"))
(define-constant +num-chatters-file+ (build-path +datadir+ "num-chatters"))

(define-constant +logdir+  (or (sys-getenv "CHATON_LOGDIR")
                               "@@server-data-dir@@logs"))

(define-constant +docdir+ (or (sys-getenv "CHATON_DOCDIR")
                              "@@server-htdocs-dir@@"))
(define-constant +status.js+ (build-path +docdir+ "var/status.js"))
(define-constant +status.scm+ (build-path +docdir+ "var/status.scm"))
(define-constant +index.rdf+ (build-path +docdir+ "var/index.rdf"))

(define-constant +show-stack-trace+
  (read-from-string "@@show-stack-trace-on-error@@"))

;;;
;;;  Entries
;;;

;; es : ((<nick> (<sec> <nsec>) <text> <ipaddr>) ...)
;; last-state : (<chatter> <ipaddr> <timestamp>)
;; renderer : optional(if omitted, chaton-render-html-1 is used)
;; returns <text-tree> and new-state
(define (chaton-render es last-state . opts)
  (let-optionals* opts ((renderer chaton-render-html-1))
    (map-accum renderer (ensure-state last-state) es)))

;; Read data file from FILE, starting from POS.
;; Returns list of entries and new POS.
(define (chaton-read-entries file pos)
  (receive (lines pos) (read-diff file pos)
    (values (safe-lines->sexps lines) pos)))

;; A convenience routine combinig above two.
;; returns <text-tree>, new-state, and new POS.
(define (chaton-render-from-file file pos last-state
                                 :key (renderer chaton-render-html-1)
                                      (newest-first #f))
  (receive (es pos) (chaton-read-entries file pos)
    (let1 es2 (if newest-first (reverse es) es)
      (receive (tree new-state) (chaton-render es2 last-state renderer)
        (values tree new-state pos)))))

;; Utility; render alist into S-expr or Json.  Assuming keys are symbols.
(define (chaton-alist->stree alist sexp?)
  (if sexp?
    (write-to-string alist)
    (letrec ([obj (lambda (x)
                    (if (list? x)
                      (array x)
                      (write-to-string x)))]
             [array (lambda (xs)
                      `("[" ,(intersperse "," (map obj xs)) "]"))])
      `("{",(intersperse "," (map (^(p)`(,(write-to-string (x->string (car p)))
                                         ":",(obj (cdr p))))
                                  alist))"}"))))

;;;
;;;  rendering
;;;

(define (ensure-state last-state) ; bridge to support backward compat. 
  (match last-state
    [(c i t) last-state]
    [_       '(#f #f #f)]))

(define (make-state chatter ip timestamp) (list chatter ip timestamp))
(define (state-chatter last-state) (car last-state))
(define (state-ip last-state)      (cadr last-state))
(define (state-timestamp last-state) (caddr last-state))

(define (chaton-render-html-1 entry last-state)
  (receive (nick sec usec text ip) (decompose-entry entry)
    (let* ([anchor-string (make-anchor-string sec usec)]
           [permalink (make-permalink sec anchor-string)])
      (values `(,(if (and (equal? nick (state-chatter last-state))
                          (equal? ip (state-ip last-state))
                          (< (abs (- (state-timestamp last-state) sec)) 240))
                   '()
                   (html:div
                    :class "entry-header"
                    (html:span :class "timestamp"
                               (sys-strftime "%Y/%m/%d %T %Z" (sys-localtime sec)))
                    (html:span :class "chatter" nick)))
                ,(html:a :class "permalink-anchor"
                         :id #`"anchor-,anchor-string"
                         :href permalink :name permalink :target "_parent"
                         "#")
                ,(html-format-entry text anchor-string))
              (make-state nick ip sec)))))

(define (chaton-render-rss-1 entry last-state)
  (receive (nick sec usec text ip) (decompose-entry entry)
    (let* ([text-with-nick #`",|nick|: ,|text|"]
           [anchor-string (make-anchor-string sec usec)]
           [permalink (make-permalink sec anchor-string)]
           [title ($ string-delete (^c (memq (char-general-category c)
                                             '(Cc Cf Cn Co Cs)))
                     $ html-escape-string (if-let1 m (#/^[^\n]*/ text-with-nick)
                                            (m 0)
                                            text-with-nick))]
           [desc (html-format-entry text-with-nick anchor-string)])
      ;; NB: DESC can never have "]]>" in it, since the external text has
      ;; gone through safe-text and all >'s in it are replaced by &gt's.
      (values `("<item>\n"
                "<title>" ,title "</title>\n"
                "<link>" ,permalink "</link>\n"
                "<description><![CDATA[" ,desc "]]></description>\n"
                "<content:encoded><![CDATA[" ,desc "]]></content:encoded>\n"
                "<pubDate>" ,(time->rfc822-date-string sec) ,"</pubDate>\n"
                "<guid isPermaLink=\"true\">" ,permalink "</guid>\n"
                "</item>\n")
              (make-state nick ip sec)))))

(define (decompose-entry entry)
  (match-let1 (nick (sec usec) text . opt) entry
    (values nick sec usec text (if (pair? opt) (car opt) #f))))

(define (make-anchor-string sec usec) (format "entry-~x-~2,'0x" sec usec))

(define (make-permalink sec anchor)
  (build-path +archive-url+
              (format "~a#~a"
                      (sys-strftime "%Y/%m/%d" (sys-gmtime sec))
                      anchor)))

(define (html-format-entry entry-text anchor-string)
  (if (#/\n/ entry-text)
    (html:pre :class "entry-multi" :id anchor-string
              (safe-text entry-text))
    (html:div :class "entry-single" :id anchor-string
              (html:span (safe-text entry-text)))))  

(define *url-rx* #/https?:\/\/(\/\/[^\/?#\s]*)?([^?#\s\"]*(\?[^#\s\"]*)?(#[^\s\"]*)?)/)

(define (safe-text text)
  (let loop ([s text] [r '()])
    (cond
     [(string-null? s) (reverse r)]
     [(*url-rx* s)
      => (^(m)
           (loop (m'after)
                 `(,(render-url (m 0)),(html-escape-string (m'before)),@r)))]
     [else (reverse (cons (html-escape-string s) r))])))

(define (render-url url)
  (rxmatch-case url
    [#/\.(?:jpg|gif|png)$/i () (render-url-image url)]
    [#/^http:\/\/(\w{2,3}\.youtube\.com)\/watch\?v=([\w-]{1,12})/ (_ host vid)
     (render-url-youtube host vid)]
    [#/^http:\/\/www\.nicovideo\.jp\/watch\/(\w{1,13})/ (_ vid)
     (render-url-nicovideo vid)]
    [else (render-url-default url)]))

(define (render-url-default url)
  (html:a :href url :rel "nofollow" :class "link-default"
          :onclick "window.open(this.href); return false;"
          (html-escape-string url)))

(define (render-url-image url)
  (html:a :href url :rel "nofollow" :class "link-image hide-while-loading"
          :onclick "window.open(this.href); return false;"
          (html:img :src url :alt url :onload "checkImageSize(this);")))

(define (render-url-youtube host vid)
  (html:object :class "youtube"
               :width "@@embed-youtube-width@@"
               :height "@@embed-youtube-height@@"
               :type "application/x-shockwave-flash"
               :data #`"http://,|host|/v/,|vid|"
               :onload "scrollToBottom();"
               (html:param :name "movie" :value #`"http://,|host|/v/,|vid|")
               (html:param :name "wmode" :value "transparent")))

(define (render-url-nicovideo vid)
  (html:iframe :width "314" :height "176"
               :src #`"http://ext.nicovideo.jp/thumb/,|vid|"
               :scrolling "no" :class "nicovideo"
               :frameborder "0"
               :onload "scrollToBottom();"
               (html:a :href #`"http://www.nicovideo.jp/watch/,|vid|"
                       (html-escape-string
                        #`"http://www.nicovideo.jp/watch/,|vid|"))))

(define (time->rfc822-date-string seconds)
  (date->string (time-utc->date (make <time> :second seconds)) "~a, ~e ~b ~Y ~X ~z"))

;;;
;;;  Reading datafile
;;;

(define *lockfile* "@@server-data-dir@@lock")

(define (%with-chaton-lock locktype opener closer thunk)
  (ensure-lockfile)
  (let1 p #f
    (unwind-protect
        (begin (set! p (opener *lockfile*))
               (sys-fcntl p F_SETLK (make <sys-flock> :type locktype))
               (thunk))
      (when p
        (sys-fcntl p F_SETLK (make <sys-flock> :type F_UNLCK))
        (closer p)))))

(define (chaton-with-shared-locking thunk)
  (%with-chaton-lock F_RDLCK open-input-file close-input-port thunk))

(define (chaton-with-exclusive-locking thunk)
  (%with-chaton-lock F_WRLCK (cut open-output-file <> :if-exists :overwrite)
                     close-output-port thunk))

(define (ensure-lockfile)
  (unless (file-exists? *lockfile*)
    (with-output-to-file *lockfile* (cut display "lockfile\n")
                         :if-does-not-exist :create :if-exists #f)))

;; Read the source file from offset START, and returns a list of
;; lines and the updated offset that points the end of the source file.
(define (read-diff source start)
  (chaton-with-shared-locking
   (cut call-with-input-file source
        (^(in) (cond [in (port-seek in start)
                         (let1 tx (port->string-list in)
                           (values tx (port-tell in)))]
                     [else (values "" 0)]))
        :if-does-not-exist #f)))

(define (safe-read line)
  (guard (e [(<read-error> e) #f]) (read-from-string line)))

(define (safe-lines->sexps lines) (filter pair? (map safe-read lines)))

;;;
;;;  Misc. Utility
;;;

;; This feature should be built-in!

(define (with-output-to-file file thunk . args)
  (if-let1 atomic (get-keyword :atomic args #f)
    (let1 tmp #`",|file|.tmp"
      (guard (e [else (sys-unlink tmp) (raise e)])
        (apply (with-module gauche with-output-to-file)
               tmp thunk (delete-keyword :atomic args))
        (sys-rename tmp file)))
    (apply (with-module gauche with-output-to-file) file thunk args)))
